README.md
# go-imagequant
[![GoDoc](https://godoc.org/github.com/InfinityTools/go-imagequant?status.svg)](https://godoc.org/github.com/InfinityTools/go-imagequant)

## About

[libimagequant](https://github.com/ImageOptim/libimagequant/) is a small, portable C library for high-quality conversion of RGBA images to 8-bit indexed-color (palette) images.

This project provides [Go](https://golang.org/) bindings for *libimagequant*. The bindings were adapted to be closer to Go code conventions. As a result, C memory management details were reduced to a single, optional, Release() call, and pixel data can be transferred by using Go's Image interface.

## Building

go-imagequant uses the system version of the static *libimagequant* library. More information about how to build libimagequant can be found on the [libimagequant project page](https://github.com/ImageOptim/libimagequant/).

go-imagequant package path is currently `github.com/InfinityTools/imagequant`. The bindings can be built via `go build`.

This package makes use of CGO, which requires a decent C compiler to be installed. However, using `go install` removes the C compiler requirement for future invocations of `go build`.

## Overview

The basic flow is:
1. Create an Attributes structure.
2. Create an Image structure from a Go `image.Image` object or a Histogram structure.
3. Perform quantization from the Image or Histogram structure (generate palette).
4. Request a paletted Go Image object.

Basic example (full code can be found in `example/example.go`):
```
package main

import "github.com/InfinityTools/go-imagequant"

func main() {
  att := imagequant.CreateAttributes()
  defer att.Release()

  // imgIn is expected to be a valid image.Image object.
  qimg := att.CreateImage(imgIn, 0.0)

  qresult, _ := att.QuantizeImage(qimg)

  // imgOut is a paletted Go image that can be directly exported using Go's image export packages.
  imgOut, _ := att.WriteRemappedImage(qresult, qimg)
}
```

Detailed function descriptions can be found in the respective Go source files.

## Documentation

For docs, see https://godoc.org/github.com/InfinityTools/go-imagequant .

## License

*go-imagequant* is released under the BSD 2-clause license. See LICENSE for more details.

*libimagequant* itself is available under under the *GPL v3 or later* license and a commercial license for non-GPL software. See *libimagequant-license.txt* for more details.


CHANGELOG.md
#### 2018-06-07 1.1.0
* Removed the need for build scripts

#### 2018-06-03 1.0.0
* Initial public release


misc.go
package imagequant
// Collection of miscellaneous functions.

import (
  "image"
  "image/color"
)

// NRGBA converts a premultiplied color back to a normalized color with each component in range [0, 255].
func NRGBA(col color.Color) (r, g, b, a byte) {
  if nrgba, ok := col.(color.NRGBA); ok {
    r, g, b, a = nrgba.R, nrgba.G, nrgba.B, nrgba.A
  } else {
    pr, pg, pb, pa := col.RGBA()
    pa >>= 8
    if pa > 0 {
      pr >>= 8
      pr *= 0xff
      pr /= pa
      pg >>= 8
      pg *= 0xff
      pg /= pa
      pb >>= 8
      pb *= 0xff
      pb /= pa
    }
    r = byte(pr)
    g = byte(pg)
    b = byte(pb)
    a = byte(pa)
  }
  return
}

// imageToBytes32 converts the given image into a 32-bit byte array.
func imageToBytes32(img image.Image) []byte {
  w := img.Bounds().Dx()
  h := img.Bounds().Dy()
  retVal := make([]byte, w*h*4)
  dofs := 0
  for y := 0; y < h; y++ {
    for x := 0; x < w; x++ {
      r, g, b, a := img.At(x, y).RGBA()
      retVal[dofs] = byte(r >> 8)
      retVal[dofs+1] = byte(g >> 8)
      retVal[dofs+2] = byte(b >> 8)
      retVal[dofs+3] = byte(a >> 8)
      dofs += 4
    }
  }
  return retVal
}

// bytesToPaletted creates a image.Paletted object out of the given palette and pixel data.
func bytesToPaletted(width, height int, palette color.Palette, pixels []byte) *image.Paletted {
  retVal := image.NewPaletted(image.Rect(0, 0, width, height), palette)
  y0, y1 := retVal.Bounds().Min.Y, retVal.Bounds().Max.Y
  sofs := 0
  dofs := y0 * retVal.Stride
  for y := y0; y < y1; y++ {
    copy(retVal.Pix[dofs:dofs+retVal.Stride], pixels[sofs:sofs+width])
    sofs += width
    dofs += retVal.Stride
  }
  return retVal
}


build.sh
#!/bin/sh

# A build script that automatically picks the right library from the subfolders in "libs".
# Use this script if you are unable or don't want to use the system library.

show_help() {
  echo "Usage $0 [options]"
  echo ""
  echo "Options:"
  echo "  --libdir path   Override library path"
  echo "  --help          This help"
  exit 0
}

# terminate(msg: string, level: int): Print "msg" and exit with "level". Default level is 0.
terminate() {
  if test $# != 0; then
    echo $1
    shift
  fi
  if test $# != 0; then
    exit $1
  else
    exit 0
  fi
}

# Checking Go compiler
if test ! $(which go); then
  terminate "Error: Go compiler not found." 1
fi

# Package-specific settings
pkgRoot="github.com/InfinityTools"
uselibdir=0

# Evaluating command line arguments...
while test $# != 0
do
  case $1 in
  --libdir)
    shift
    if test $# = 0; then
      terminate "Missing argument: --libdir" 1
    fi
    uselibdir=1
    libdir="$1"
    ;;
  --help)
    show_help
    ;;
  esac
  shift
done

# Setting package-specific linker options
ldargs="-limagequant -lm"

if test $uselibdir = 0; then
  libos=$(go env GOOS)
  libarch=$(go env GOARCH)
  echo "Detected: os=$libos, arch=$libarch"

  pkgImagequant=$pkgRoot/go-imagequant
  ldprefix=$(go list -f {{.Dir}} $pkgImagequant)
  test $? = 0 || terminate "Package not found: $pkgImagequant" 1
  libdir="$ldprefix/libs/$libos/$libarch"
else
    echo "Using libdir: $libdir"
fi

if test ! -d "$libdir"; then
  terminate "Error: Path does not exist: $libdir" 1
fi

echo "Building library..."
CGO_LDFLAGS="-L$libdir $ldargs" go build && go install && terminate "Finished." 0 || terminate "Cancelled." 1


image.go
package imagequant

/*
#include "libimagequant.h"
*/
import "C"

import (
  "image"
  "image/color"
  "runtime"
  "unsafe"
)


// Image struct is required by several functions. Don't access the content directly.
type Image struct {
  image     *C.struct_liq_image
  buffer      []byte    // set to prevent GC from cleaning up pixel buffer prematurely
  bufferRows  [][]byte  // set to prevent GC from cleaning up pixel buffer prematurely
}


// Creates an object that represents the image pixels to be used for quantization and remapping.
//
// The pixel array must be contiguous run of RGBA pixels (alpha is the last component, 0 = transparent, 255 = opaque).
//
// The rgba array must not be modified or freed until this object is freed with liq_image_destroy. See also liq_image_set_memory_ownership.
//
// width and height are dimensions in pixels. An image 10x10 pixel large will need a 400-byte array.
//
// gamma can be 0 for images with the typical 1/2.2 gamma. Otherwise gamma must be > 0 and < 1, e.g. 0.45455 (1/2.2) or 0.55555 (1/1.8).
//
// Generated palette will use the same gamma unless SetOutputGamma is used. If SetOutputGamma is not used, then it only affects whether
// brighter or darker areas of the image will get more palette colors allocated.
//
// Returns nil on failure, e.g. if rgba is nil or too small or width/height is <= 0.
func (att *Attributes) CreateImageBuffer(rgba []byte, width, height int, gamma float64) *Image {
  if width < 0 || height < 0 { return nil }
  if rgba == nil || len(rgba) < width*height*4 { return nil }
  // img := Image{ nil }
  img := new(Image)
  img.image = C.liq_image_create_rgba(att.attr, unsafe.Pointer(&rgba[0]), C.int(width), C.int(height), C.double(gamma))
  if img.image == nil { return nil }
  img.buffer = rgba
  runtime.SetFinalizer(img, freeImage)
  return img
}

// Same as CreateImageBuffer, but takes an array of rows of pixels.
//
// This allows defining images with reversed rows (like in BMP), "stride" different than width or using only fragment of a larger bitmap, etc.
// The rows array must have at least height elements, and each row must be at least width RGBA pixels wide.
func (att *Attributes) CreateImageBufferRows(rgbaRows [][]byte, width, height int, gamma float64) *Image {
  if width < 0 || height < 0 { return nil }
  if rgbaRows == nil || len(rgbaRows) < height { return nil }

  // img := Image{ nil }
  img := new(Image)
  rowPtr := make([]uintptr, len(rgbaRows))
  for i := 0; i < len(rgbaRows); i++ {
    if rgbaRows[i] == nil || len(rgbaRows[i]) < width * 4 { return nil }
    rowPtr[i] = uintptr(unsafe.Pointer(&rgbaRows[i][0]))
  }
  img.image = C.liq_image_create_rgba_rows(att.attr, (*unsafe.Pointer)(unsafe.Pointer(&rowPtr[0])), C.int(width), C.int(height), C.double(gamma))
  if img.image == nil { return nil }
  img.bufferRows = rgbaRows
  runtime.SetFinalizer(img, freeImage)
  return img
}

// Same as CreateImageBuffer, but takes a Go Image interface as source.
func (att *Attributes) CreateImage(img image.Image, gamma float64) *Image {
  buf := imageToBytes32(img)
  width, height := img.Bounds().Dx(), img.Bounds().Dy()
  return att.CreateImageBuffer(buf, width, height, gamma)
}

// Analyze and remap this image with assumption that it will be always presented exactly on top of this background.
//
// When this image is remapped to a palette with a fully transparent color (use AddImageFixedColor to ensure this)
// pixels that are better represented by the background than the palette will be made transparent. This function can
// be used to improve quality of animated GIFs by setting previous animation frame as the background.
//
// Returns ErrBufferTooSmall if the background image has a different size than the foreground.
func (att *Attributes) SetImageBackground(img *Image, background *Image) error {
  code := C.liq_image_set_background(img.image, background.image)
  return getError(code)
}

// Importance map controls which areas of the image get more palette colors.
//
// Pixels corresponding to 0 values in the map are completely ignored.
// The higher the value the more weight is placed on the given pixel, giving it higher chance of influencing the final palette.
// The map is one byte per pixel and must have the same size as the image (width×height bytes).
//
// Returns ErrInvalidPointer if any pointer is nil and ErrBufferTooSmall if the map size does not match the image size.
func (att *Attributes) SetImageImportanceMap(img *Image, importanceMap []byte) error {
  code := C.liq_image_set_importance_map(img.image, (*C.uchar)(unsafe.Pointer(&importanceMap[0])), C.size_t(len(importanceMap)), C.LIQ_COPY_PIXELS)
  return getError(code)
}

// Reserves a color in the output palette created from this image.
//
// It behaves as if the given color was used in the image and was very important.
// RGB values of the Color object are assumed to have the same gamma as the image. It must be called before the image is quantized.
//
// Returns error if more than 256 colors are added. If image is quantized to fewer colors than the number of fixed colors added, then excess fixed colors will be ignored.
func (att *Attributes) AddImageFixedColor(img *Image, col color.Color) error {
  r, g, b, a := col.RGBA()
  c := C.struct_liq_color{}
  c.r, c.g, c.b, c.a = C.uchar(r), C.uchar(g), C.uchar(b), C.uchar(a)
  code := C.liq_image_add_fixed_color(img.image, c)
  return getError(code)
}

// Getter for image width.
func (att *Attributes) GetImageWidth(img *Image) int {
  return int(C.liq_image_get_width(img.image))
}

// Getter for image height.
func (att *Attributes) GetImageHeight(img *Image) int {
  return int(C.liq_image_get_height(img.image))
}


// Used internally. Frees an Image object.
func freeImage(i *Image) {
  if i.image != nil {
    // fmt.Println("Releasing Image object.")
    C.liq_image_destroy(i.image)
    i.image = nil
    i.buffer = nil
    i.bufferRows = nil
  }
}


histogram.go
package imagequant

/*
#include "libimagequant.h"
*/
import "C"

import (
  "image/color"
  "runtime"
  "unsafe"
)


// A HistogramEntry holds usage information of a single color value.
type HistogramEntry struct {
  Color     color.Color // The color value definition
  Count     uint        // Number of occurrence, influences the weight or importance of the color.
}

// Histogram struct is required by several functions. Don't accss the content directly.
type Histogram struct {
  histogram *C.struct_liq_histogram
}


// Creates a histogram object that will be used to collect color statistics from multiple images.
func (att *Attributes) CreateHistogram() *Histogram {
  hist := new(Histogram)
  hist.histogram = C.liq_histogram_create(att.attr)
  runtime.SetFinalizer(hist, freeHistogram)
  return hist
}

// "Learns" colors from the image, which will be later used to generate the palette.
//
// After the image is added to the histogram it may be freed to save memory (but it's more efficient to keep the image object around if it's going to be used for remapping).
// Fixed colors added to the image are also added to the histogram. If total number of fixed colors exceeds 256, this function will fail with ErrBufferTooSmall.
func (att *Attributes) AddImageToHistogram(hist *Histogram, img *Image) error {
  code := C.liq_histogram_add_image(hist.histogram, att.attr, img.image)
  return getError(code)
}

// Alternative to AddImageToHistogram. Instead of counting colors in an image, it directly takes an array of colors and their counts.
//
// This function is only useful if you already have a histogram of the image from another source.
func (att *Attributes) AddColorsToHistogram(hist *Histogram, entries []HistogramEntry, gamma float64) error {
  if entries == nil { return ErrInvalidPointer }
  c_entries := make([]C.struct_liq_histogram_entry, len(entries))
  for k, v := range entries {
    r, g, b, a := v.Color.RGBA()
    c_entries[k].color.r = C.uchar(r)
    c_entries[k].color.g = C.uchar(g)
    c_entries[k].color.b = C.uchar(b)
    c_entries[k].color.a = C.uchar(a)
    c_entries[k].count = C.uint(v.Count)
  }
  code := C.liq_histogram_add_colors(hist.histogram, att.attr,
                                     (*C.struct_liq_histogram_entry)(unsafe.Pointer(&c_entries[0])),
                                     C.int(len(c_entries)), C.double(gamma))
  return getError(code)
}


// Used internally. Frees a Histogram object.
func freeHistogram(h *Histogram) {
  if h.histogram != nil {
    // fmt.Println("Releasing Histogram object.")
    C.liq_histogram_destroy(h.histogram)
    h.histogram = nil
  }
}


attributes.go
package imagequant

/*
#include "libimagequant.h"
*/
import "C"

import (
  "runtime"
)

const (
  SPEED_SLOWEST = 1
  SPEED_DEFAULT = 3
  SPEED_FASTEST = 10
)

const (
  QUALITY_BEST  = 100
  QUALITY_GOOD  = 80
  QUALITY_WORST = 0
)

// The Attributes struct is used to call the majority of quantization functions.
// This is the only structure that can be released manually.
type Attributes struct {
  attr      *C.struct_liq_attr
}


// Returns an object that will hold initial settings (attributes) for the library.
//
// IMPORTANT: The object must be freed by Release after it is no longer needed.
func CreateAttributes() *Attributes {
  att := new(Attributes)
  att.attr = C.liq_attr_create()
  runtime.SetFinalizer(att, freeAttribute)
  return att
}

// Creates an independent copy of the calling object.
//
// IMPORTANT: The copy must also be freed by the Release function.
func (att *Attributes) CopyAttribute() *Attributes {
  att2 := new(Attributes)
  att2.attr = C.liq_attr_copy(att.attr)
  runtime.SetFinalizer(att2, freeAttribute)
  return att2
}

// Call this function to manually release the attributes.
// Otherwise, Golang's own garbage collector will take care of it eventually.
func (att *Attributes) Release() {
  freeAttribute(att)
}


// Specifies the maximum number of colors to use. The default is 256.
//
// Instead of setting a fixed limit it's better to use SetQuality.
// Returns ErrValueOutOfRange if number of colors is outside the range 2-256.
func (att *Attributes) SetMaxColors(colors int) error {
  code := C.liq_set_max_colors(att.attr, C.int(colors))
  return getError(code)
}

// Returns the value set by SetMaxColors.
func (att *Attributes) GetMaxColors() int {
  retVal := C.liq_get_max_colors(att.attr)
  return int(retVal)
}

// Higher speed levels disable expensive algorithms and reduce quantization precision. The default speed is 3.
//
// Speed 1 gives marginally better quality at significant CPU cost. Speed 10 has usually 5% lower quality, but is 8 times faster than the default.
// High speeds combined with SetQuality will use more colors than necessary and will be less likely to meet minimum required quality.
// The range of SPEED_xxx constants covers the common uses.
// Features dependent on speed:
//    Noise-sensitive dithering speed     1 to 5
//    Forced posterization                8-10 or if image has more than million colors
//    Quantization error known            1-7 or if minimum quality is set
//    Additional quantization techniques  1-6
// Returns ErrValueOutOfRange if the speed is outside the 1-10 range.
func (att *Attributes) SetSpeed(speed int) error {
  code := C.liq_set_speed(att.attr, C.int(speed))
  return getError(code)
}

// Returns the value set by SetSpeed.
func (att *Attributes) GetSpeed() int {
  retVal := C.liq_get_speed(att.attr)
  return int(retVal)
}

// Ignores the given number of least significant bits in all channels, posterizing image to 2^bits levels.
//
// 0 gives full quality. Use 2 for VGA or 16-bit RGB565 displays, 4 if image is going to be output
// on a RGB444/RGBA4444 display (e.g. low-quality textures on Android).
//
// Returns LIQ_VALUE_OUT_OF_RANGE if the value is outside the 0-4 range.
func (att *Attributes) SetMinPosterization(bits int) error {
  code := C.liq_set_min_posterization(att.attr, C.int(bits))
  return getError(code)
}

// Returns the value set by SetMinPosterization.
func (att *Attributes) GetMinPosterization() int {
  retVal := C.liq_get_min_posterization(att.attr)
  return int(retVal)
}

// Quality is in range 0 (worst) to 100 (best) and values are analoguous to JPEG quality (i.e. 80 is usually good enough).
//
// Quantization will attempt to use the lowest number of colors needed to achieve maximum quality. max value of 100 is the default
// and means conversion as good as possible. If it's not possible to convert the image with at least minimum quality (i.e. 256 colors
// is not enough to meet the minimum quality), then liq_image_quantize will fail. The default minimum is 0 (proceeds regardless of quality).
//
// Quality measures how well the generated palette fits image given to liq_image_quantize. If a different image is remapped with
// WriteRemappedImage, then actual quality may be different.
// Regardless of the quality settings the number of colors won't exceed the maximum (see SetMaxColors).
// The range of QUALITY_xxx constants covers the common uses.
//
// Returns ErrValueOutOfRange if target is lower than minimum or any of them is outside the 0-100 range.
// Returns ErrInvalidPointer if attr appears to be invalid.
func (att *Attributes) SetQuality(min, max int) error {
  code := C.liq_set_quality(att.attr, C.int(min), C.int(max))
  return getError(code)
}

// Returns the minimum/maximum range of quality set by SetQuality.
func (att *Attributes) GetQuality() (min, max int) {
  min = int(C.liq_get_min_quality(att.attr))
  max = int(C.liq_get_max_quality(att.attr))
  return
}

// Setting to false makes alpha colors sorted before opaque colors. "true" mixes colors together except completely transparent color,
// which is moved to the end of the palette. This is a workaround for programs that blindly assume the last palette entry is transparent.
func (att *Attributes) SetLastIndexTransparent(set bool) {
  v := 0
  if set { v = 1 }
  C.liq_set_last_index_transparent(att.attr, C.int(v))
}


// Used internally. Frees a Attributes object.
func freeAttribute(att *Attributes) {
  if att.attr != nil {
    // fmt.Println("Releasing Attributes object.")
    C.liq_attr_destroy(att.attr)
    att.attr = nil
    runtime.GC()
  }
}


imagequant.go
/*
Package imagequant provides bindings to the external imagequant C library.

Original C library: https://github.com/ImageOptim/libimagequant/
*/
package imagequant
// Alternative Go binding package (by larrabee): https://github.com/ultimate-guitar/go-imagequant
// TODO: Attempt to implement callback support:
// - liq_set_log_callback
// - liq_set_log_flush_callback
// - liq_attr_set_progress_callback
// - liq_result_set_progress_callback
// - liq_image_create_custom

/*
// CGO linker flags are defined for selected platforms windows/linux/freebsd/darwin and architectures 386/amd64.
// Set CGO_LDFLAGS environment variable manually for undefined platforms and architectures or non-standard configurations.
// Note: Replace uncommented with commented Windows cgo directives to enable static linking.
// #cgo windows,386 LDFLAGS: -Llibs/windows/386 -limagequant -lm -static
// #cgo windows,amd64 LDFLAGS: -Llibs/windows/amd64 -limagequant -lm -static
#cgo windows,386 LDFLAGS: -Llibs/windows/386 -limagequant -lm
#cgo windows,amd64 LDFLAGS: -Llibs/windows/amd64 -limagequant -lm
#cgo linux,386 LDFLAGS: -Llibs/linux/386 -limagequant -lm
#cgo linux,amd64 LDFLAGS: -Llibs/linux/amd64 -limagequant -lm
#cgo freebsd,386 LDFLAGS: -Llibs/freebsd/386 -limagequant -lm
#cgo freebsd,amd64 LDFLAGS: -Llibs/freebsd/amd64 -limagequant -lm
#cgo darwin,386 LDFLAGS: -Llibs/darwin/386 -limagequant -lm
#cgo darwin,amd64 LDFLAGS: -Llibs/darwin/amd64 -limagequant -lm
#include "libimagequant.h"

const char* liqVersionString() {
  return LIQ_VERSION_STRING;
}

*/
import "C"

import (
  "errors"
)


var (
  // Potential error codes
  ErrQualityTooLow      = errors.New("Quality too low")
  ErrValueOutOfRange    = errors.New("Value is out of range")
  ErrOutOfMemory        = errors.New("Out of memory")
  ErrAborted            = errors.New("Aborted")
  ErrBitmapNotAvailable = errors.New("Bitmap is not available")
  ErrBufferTooSmall     = errors.New("Buffer is too small")
  ErrInvalidPointer     = errors.New("Invalid pointer")
  ErrUnsupported        = errors.New("Unsupported")
  ErrUnknown            = errors.New("Unknown error")
)


// GetVersion returns the imagequant library version as major, minor and patch number.
func GetVersion() (major, minor, patch int) {
  value := int(C.liq_version())
  patch = value % 100
  value /= 100
  minor = value % 100
  value /= 100
  major = value
  return
}

// GetVersionString returns the imagequant library version as a preformatted string.
func GetVersionString() string {
  return C.GoString(C.liqVersionString())
}


// Used internally. Translates error codes to Golang errors.
func getError(code C.liq_error) error {
  switch code {
  case C.LIQ_OK:
    return nil
  case C.LIQ_QUALITY_TOO_LOW:
    return ErrQualityTooLow
  case C.LIQ_VALUE_OUT_OF_RANGE:
    return ErrValueOutOfRange
  case C.LIQ_OUT_OF_MEMORY:
    return ErrOutOfMemory
  case C.LIQ_ABORTED:
    return ErrAborted
  case C.LIQ_BITMAP_NOT_AVAILABLE:
    return ErrBitmapNotAvailable
  case C.LIQ_BUFFER_TOO_SMALL:
    return ErrBufferTooSmall
  case C.LIQ_INVALID_POINTER:
    return ErrInvalidPointer
  case C.LIQ_UNSUPPORTED:
    return ErrUnsupported
  default:
    return ErrUnknown
  }
}


example/example.go
package main

import (
  "flag"
  "fmt"
  "image/png"
  "os"

  "github.com/InfinityTools/go-imagequant"
)

func main() {
  ShouldDisplayVersion := flag.Bool("Version", false, "Print library version and exit")
  InFile := flag.String("In", "", "Input PNG filename")
  OutFile := flag.String("Out", "", "Output PNG filename")
  Speed := flag.Int("Speed", 3, "Speed (1 slowest, 10 fastest)")
  Compression := flag.Int("Compression", -3, "Compression level (DefaultCompression = 0, NoCompression = -1, BestSpeed = -2, BestCompression = -3)")

  flag.Parse()
  if len(os.Args) < 2 {
    fmt.Printf("Usage: %s [options]\nOptions:\n", os.Args[0])
    flag.PrintDefaults()
    os.Exit(0)
  }

  if *ShouldDisplayVersion {
    fmt.Printf("libimagequant '%s'\n", imagequant.GetVersionString())
    os.Exit(0)
  }

  var cLevel png.CompressionLevel
  switch *Compression {
    case 0:
      cLevel = png.DefaultCompression
    case -1:
      cLevel = png.NoCompression
    case -2:
      cLevel = png.BestSpeed
    case -3:
      cLevel = png.BestCompression
    default:
      cLevel = png.BestCompression
  }

  err := quantizeFile(*InFile, *OutFile, *Speed, cLevel)
  if err != nil {
    fmt.Println(err.Error())
    os.Exit(1)
  }

  os.Exit(0)
}

func quantizeFile(inFile, outFile string, speed int, compression png.CompressionLevel) error {
  fin, err := os.OpenFile(inFile, os.O_RDONLY, 0444)
  if err != nil {
    return fmt.Errorf("os.OpenFile: %s", err.Error())
  }
  defer fin.Close()

  imgIn, err := png.Decode(fin)
  if err != nil {
    return fmt.Errorf("png.Decode: %s", err.Error())
  }

  // This function call initializes the main quantization structure.
  // Call it once before starting the quantization process.
  quant := imagequant.CreateAttributes()
  // It is recommended but not required to release this object after use. Go's garbage collector
  // will eventually release it automatically when it is no longer used.
  defer quant.Release()

  // Add the source image to the quantizer. Second argument "gamma" is left to the default 0.
  qimg := quant.CreateImage(imgIn, 0.0)
  if qimg == nil {
    return fmt.Errorf("quant.CreateImage: Could not initialize image structure")
  }

  // Generate the palette.
  qresult, err := quant.QuantizeImage(qimg)
  if err != nil {
    return fmt.Errorf("quant.QuantizeImage: %s", err.Error())
  }

  // Remap the input image to the palette and return it as paletted image.
  imgOut, err := quant.WriteRemappedImage(qresult, qimg)
  if err != nil {
    return fmt.Errorf("quant.WriteRemappedImage: %s", err.Error())
  }

  fout, err := os.OpenFile(outFile, os.O_WRONLY|os.O_CREATE, 0644)
  if err != nil {
    return fmt.Errorf("os.OpenFile: %s", err.Error())
  }
  defer fout.Close()

  encoder := png.Encoder{CompressionLevel: compression, BufferPool: nil}
  err = encoder.Encode(fout, imgOut)
  if err != nil {
    return fmt.Errorf("png.Encode: %s", err.Error())
  }

  return nil
}


result.go
package imagequant

/*
#include "libimagequant.h"
*/
import "C"

import (
  "image"
  "image/color"
  "runtime"
  "unsafe"
)


const (
  DITHER_MIN  = 0.0
  DITHER_MAX  = 1.0
)

// Result struct is required by several functions. Don't access the content directly.
type Result struct {
  result    *C.struct_liq_result
}


// Generates a palette from the histogram. On success returns the fully initialized Result object.
func (att *Attributes) QuantizeHistogram(hist *Histogram) (res *Result, err error) {
  res = new(Result)
  code := C.liq_histogram_quantize(hist.histogram, att.attr, (**C.struct_liq_result)(unsafe.Pointer(&res.result)))
  runtime.SetFinalizer(res, freeResult)
  err = getError(code)
  return
}

// Performs quantization (palette generation) based on current Quantizer settings and pixels of the image.
//
// Returns the Result object if quantization succeeds.
// Error returns ErrQualityTooLow if quantization fails due to limit set in SetQuality.
func (att *Attributes) QuantizeImage(img *Image) (res *Result, err error) {
  res = new(Result)
  code := C.liq_image_quantize(img.image, att.attr, (**C.struct_liq_result)(unsafe.Pointer(&res.result)))
  runtime.SetFinalizer(res, freeResult)
  err = getError(code)
  return
}

// Enables/disables dithering in WriteRemappedImage.
//
// Dithering level must be between 0 and 1 (inclusive). Dithering level 0 enables fast non-dithered remapping.
// Otherwise a variation of Floyd-Steinberg error diffusion is used.
func (att *Attributes) SetDitheringLevel(res *Result, ditherLevel float32) error {
  code := C.liq_set_dithering_level(res.result, C.float(ditherLevel))
  return getError(code)
}

// Sets gamma correction for generated palette and remapped image.
//
// Must be > 0 and < 1, e.g. 0.45455 for gamma 1/2.2 in PNG images. By default output gamma is same as gamma of the input image.
func (att *Attributes) SetOutputGamma(res *Result, gamma float64) error {
  code := C.liq_set_output_gamma(res.result, C.double(gamma))
  return getError(code)
}

// Returns the gamma value for the output image.
func (att *Attributes) GetOutputGamma(res *Result) float64 {
  return float64(C.liq_get_output_gamma(res.result))
}

// Returns a palette optimized for the image that has been quantized or remapped (final refinements are applied to the palette during remapping).
//
// It's valid to call this method before remapping, if you don't plan to remap any images or want to use same palette for multiple images.
// Returns a Palette object with 0 color entries on error.
func (att *Attributes) GetPalette(res *Result) color.Palette {
  var palette color.Palette = nil
  pal := C.liq_get_palette(res.result)
  if pal != nil {
    palette = make(color.Palette, (*pal).count)
    for i := 0; i < len(palette); i++ {
      r, g, b, a := byte((*pal).entries[i].r), byte((*pal).entries[i].g), byte((*pal).entries[i].b), byte((*pal).entries[i].a)
      // compensating rounding errors
      if r > a { r = a }
      if g > a { g = a }
      if b > a { b = a }
      palette[i] = color.RGBA{ r, g, b, a }
    }
  } else {
    palette = make(color.Palette, 0)
  }
  return palette
}

// Remaps the image to palette and returns the converted image as a byte array, 1 pixel per byte.
//
// For best performance call GetPalette after this function, as palette is improved during remapping
// (except when QuantizeHistogram is used).
//
// The returned byte array is assumed to be contiguous, with rows ordered from top to bottom, and no gaps between rows.
// If you need to return a sequence of rows with padding or upside-down order, then use WriteRemappedImageRows.
func (att *Attributes) WriteRemappedImageBuffer(res *Result, img *Image) (buf []byte, err error) {
  buf = make([]byte, att.GetImageWidth(img) * att.GetImageHeight(img))
  code := C.liq_write_remapped_image(res.result, img.image, unsafe.Pointer(&buf[0]), C.size_t(len(buf)))
  err = getError(code)
  return
}

// Similar to WriteRemappedImageBuffer. Returns a remapped image, at 1 byte per pixel, to each row pointed by rows multi-array.
//
// The array must have at least as many elements as height of the image, and each row must have at least as many bytes as width of the image.
// Rows must not overlap.
func (att *Attributes) WriteRemappedImageBufferRows(res *Result, img *Image, rows [][]byte) (rowsOut [][]byte, err error) {
  if rows == nil { err = ErrInvalidPointer; return }
  if len(rows) < att.GetImageHeight(img) { err = ErrBufferTooSmall; return }

  rowPtr := make([]*C.uchar, len(rows))
  width := att.GetImageWidth(img)
  for i := 0; i < len(rows); i++ {
    if rows[i] == nil || len(rows[i]) < width { err = ErrBufferTooSmall; return }
    rowPtr[i] = (*C.uchar)(unsafe.Pointer(&rows[i][0]))
  }
  code := C.liq_write_remapped_image_rows(res.result, img.image, (**C.uchar)(unsafe.Pointer(&rowPtr[0])))
  rowsOut = rows
  err = getError(code)
  return
}

// A convenience function that returns a paletted Go Image object.
func (att *Attributes) WriteRemappedImage(res *Result, img *Image) (imgOut image.Image, err error) {
  buf, err := att.WriteRemappedImageBuffer(res, img)
  if err != nil { return }

  pal := att.GetPalette(res)
  if len(pal) == 0 { err = ErrUnknown; return }

  width, height := att.GetImageWidth(img), att.GetImageHeight(img)
  imgOut = bytesToPaletted(width, height, pal, buf)
  return
}


// Returns mean square error of quantization (square of difference between pixel values in the source image and its remapped version).
//
// Alpha channel, gamma correction and approximate importance of pixels is taken into account, so the result isn't exactly the mean square error of all channels.
// or most images MSE 1-5 is excellent. 7-10 is OK. 20-30 will have noticeable errors. 100 is awful.
//
// This function may return -1 if the value is not available (this happens when a high speed has been requested, the image hasn't been remapped yet,
// and quality limit hasn't been set, see SetSpeed and SetQuality). The value is not updated when multiple images are remapped, it applies only to the image
// used in QuantizeImage or the first image that has been remapped. See GetRemappingError.
func (att *Attributes) GetQuantizationError(res *Result) float64 {
  return float64(C.liq_get_quantization_error(res.result))
}

// Analoguous to GetQuantizationError, but returns quantization error as quality value in the same 0-100 range that is used by SetQuality.
//
// It may return -1 if the value is not available (see note in GetQuantizationError).
func (att *Attributes) GetQuantizationQuality(res *Result) int {
  return int(C.liq_get_quantization_quality(res.result))
}

// Returns mean square error of last remapping done (square of difference between pixel values in the remapped image and its remapped version).
//
// Alpha channel and gamma correction are taken into account, so the result isn't exactly the mean square error of all channels.
func (att *Attributes) GetRemappingError(res *Result) float64 {
  return float64(C.liq_get_remapping_error(res.result))
}

// Analoguous to GetRemappingError, but returns quantization error as quality value in the same 0-100 range that is used by SetQuality.
func (att *Attributes) GetRemappingQuality(res *Result) int {
  return int(C.liq_get_remapping_quality(res.result))
}


// Used internally. Frees a Result object.
func freeResult(r *Result) {
  if r.result != nil {
    // fmt.Println("Releasing Result object.")
    C.liq_result_destroy(r.result)
    r.result = nil
  }
}
